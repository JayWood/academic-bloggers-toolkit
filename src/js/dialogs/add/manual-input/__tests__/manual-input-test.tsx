jest.mock('utils/resolvers/url');
jest.mock('utils/resolvers/isbn');
import { mount, ReactWrapper, ShallowWrapper, shallow } from 'enzyme';
import toJSON, { OutputMapper } from 'enzyme-to-json';
import React from 'react';

import Store from 'stores/ui/add-dialog';
import ManualInput from '..';

const setup = (citationType: CSL.ItemType = 'webpage', fullMount = true) => {
    const store = new Store(citationType);
    const component = fullMount
        ? mount(<ManualInput store={store} />)
        : shallow(<ManualInput store={store} />);
    return {
        component,
        instance: component.instance() as ManualInput,
    };
};

const mapper: OutputMapper = json => {
    if (json.props.meta) {
        json.props.meta = '__SKIPPED__';
    }
    if (json.props.store) {
        json.props.store = '__SKIPPED__';
    }
    if (json.props.id) {
        json.props.id = '__SKIPPED__';
    }
    return json;
};

const J = (component: ReactWrapper | ShallowWrapper) =>
    toJSON(component, { noKey: true, map: mapper });

describe('<ManualInput />', () => {
    const BASELINE = J(setup().component);
    it('should match baseline snapshot', () => {
        expect(BASELINE).toMatchSnapshot();
    });
    it('should render autocite fields for books', () => {
        const { component } = setup('book');
        expect(J(component)).toMatchDiffSnapshot(BASELINE);
    });
    it('should not render autocite for anything but websites or books', () => {
        const { component } = setup('article');
        expect(J(component)).toMatchDiffSnapshot(BASELINE);
    });
    it('should handle url autocites', async () => {
        expect.assertions(1);
        const { instance } = setup();
        const before = instance.props.store.data.CSL;
        await instance.handleAutocite('https://google.com');
        const after = instance.props.store.data.CSL;
        delete before.id;
        delete after.id;
        expect(before).toMatchDiffSnapshot(after);
    });
    it('should handle isbn autocites', async () => {
        expect.assertions(1);
        const { instance } = setup('book');
        const before = instance.props.store.data.CSL;
        await instance.handleAutocite('1234567890');
        const after = instance.props.store.data.CSL;
        delete before.id;
        delete after.id;
        expect(before).toMatchDiffSnapshot(after);
    });
    it('should gracefully handle errors', async () => {
        expect.assertions(4);
        const { instance } = setup();
        const before = instance.props.store.data.CSL;
        expect(instance.props.store.errorMessage).toBe('');
        await instance.handleAutocite('ERROR');
        const after = instance.props.store.data.CSL;
        expect(before).toEqual(after);
        expect(instance.props.store.errorMessage).toBe('Testing errors');
        instance.setErrorMessage();
        instance.focusTypeSelect(null);
        expect(instance.props.store.errorMessage).toBe('');
    });
    it('should handle type changes', () => {
        const { instance } = setup();
        expect(instance.props.store.data.citationType).toBe('webpage');
        instance.handleTypeChange({ currentTarget: { value: 'book' } } as any);
        expect(instance.props.store.data.citationType).toBe('book');
    });
    describe('wheel event tests', () => {
        const { component } = setup('webpage', false);
        const wheelDiv = component.find('.scrollBoundaryAutocite');
        let atTop: any = {};
        let inMiddle: any = {};
        let atBottom: any = {};
        beforeEach(() => {
            atTop = {
                cancelable: true,
                deltaY: 0,
                currentTarget: {
                    clientHeight: 50,
                    scrollHeight: 100,
                    scrollTop: 0,
                },
            };
            inMiddle = {
                cancelable: true,
                deltaY: 0,
                currentTarget: {
                    clientHeight: 50,
                    scrollHeight: 100,
                    scrollTop: 25,
                },
            };
            atBottom = {
                cancelable: true,
                deltaY: 0,
                currentTarget: {
                    clientHeight: 50,
                    scrollHeight: 100,
                    scrollTop: 50,
                },
            };
        });
        it('should cancel at top scrolling up', () => {
            atTop.deltaY = -5;
            wheelDiv.simulate('wheel', atTop);
            expect(atTop.cancelable).toBe(true);
        });
        it('should cancel at bottom scrolling down', () => {
            atBottom.deltaY = 5;
            wheelDiv.simulate('wheel', atBottom);
            expect(atBottom.cancelable).toBe(true);
        });
        it('should NOT cancel in middle scrolling up', () => {
            inMiddle.deltaY = -5;
            wheelDiv.simulate('wheel', inMiddle);
            expect(inMiddle.cancelable).toBe(false);
        });
        it('should NOT cancel in middle scrolling down', () => {
            inMiddle.deltaY = 5;
            wheelDiv.simulate('wheel', inMiddle);
            expect(inMiddle.cancelable).toBe(false);
        });
    });
});

// const setup = ({ itemType = 'webpage', fullMount = true } = {}) => {
//     manualData.replace(new Map([['type', itemType]]));
//     const component = fullMount
//         ? mount(
//               <ManualEntryContainer
//                   manualData={manualData}
//                   errorMessage={errorMessage}
//                   people={people}
//                   {...mocks}
//               />,
//           )
//         : shallow(
//               <ManualEntryContainer
//                   manualData={manualData}
//                   errorMessage={errorMessage}
//                   people={people}
//                   {...mocks}
//               />,
//           );
//     return {
//         component,
//         instance: component.instance() as ManualEntryContainer,
//     };
// };

// describe('<ManualEntryContainer />', () => {
//     beforeEach(() => {
//         people.replace([{ ...manualPersonObj }]);
//         errorMessage.set('');
//         jest.resetAllMocks();
//     });
//     it('should match snapshots', () => {
//         let { component } = setup();
//         expect(toJSON(component, { noKey: true })).toMatchSnapshot();

//         ({ component } = setup({ itemType: 'book' }));
//         expect(toJSON(component, { noKey: true })).toMatchSnapshot();

//         ({ component } = setup({ itemType: 'article' }));
//         expect(toJSON(component, { noKey: true })).toMatchSnapshot();
//     });
//     it('should show and hide error callout', () => {
//         errorMessage.set('Hello World');
//         const { component, instance } = setup();
//         expect(component.find('Callout').children().length).toBeGreaterThan(0);
//         instance.dismissError();
//         component.update();
//         expect(component.find('Callout').children().length).toBe(0);
//     });
//     it('should handle type change', () => {
//         const { component } = setup();
//         const select = component.find('#type-select');
//         select.simulate('change', { currentTarget: { value: 'book' } });
//         expect(mocks.onTypeChange).toHaveBeenCalledTimes(1);
//     });
//     it('should not break when ref callback receives null', () => {
//         const { instance } = setup();
//         expect(() => instance.focusTypeSelect(null)).not.toThrow();
//     });
// });
