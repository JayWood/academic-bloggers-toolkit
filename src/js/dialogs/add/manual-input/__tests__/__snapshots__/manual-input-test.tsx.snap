// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManualInput /> should handle isbn autocites 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
+   \\"ISBN\\": \\"1234567890\\",
+   \\"author\\": Array [
+     Object {
+       \\"family\\": \\"Doe\\",
+       \\"given\\": \\"John\\",
+       \\"literal\\": \\"\\",
+     },
+     Object {
+       \\"family\\": \\"Smith\\",
+       \\"given\\": \\"Jane\\",
+       \\"literal\\": \\"\\",
+     },
+   ],
+   \\"issued\\": Object {
+     \\"date-parts\\": Array [
+       Array [
+         \\"2017\\",
+         \\"01\\",
+         \\"01\\",
+       ],
+     ],
+   },
+   \\"number-of-pages\\": \\"100\\",
+   \\"publisher\\": \\"Mocked Publisher\\",
+   \\"title\\": \\"Mocked Response Title\\",
    \\"type\\": \\"book\\",
  }"
`;

exports[`<ManualInput /> should handle url autocites 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
+   \\"URL\\": \\"https://google.com\\",
+   \\"accessed\\": Object {
+     \\"date-parts\\": Array [
+       Array [
+         \\"2018\\",
+         \\"01\\",
+         \\"01\\",
+       ],
+     ],
+   },
+   \\"author\\": Array [
+     Object {
+       \\"family\\": \\"Doe\\",
+       \\"given\\": \\"John\\",
+       \\"literal\\": \\"\\",
+     },
+     Object {
+       \\"family\\": \\"Smith\\",
+       \\"given\\": \\"Jane\\",
+       \\"literal\\": \\"\\",
+     },
+   ],
+   \\"container-title\\": \\"Mocked Site Title\\",
+   \\"issued\\": Object {
+     \\"date-parts\\": Array [
+       Array [
+         \\"2017\\",
+         \\"01\\",
+         \\"01\\",
+       ],
+     ],
+   },
+   \\"title\\": \\"Mocked Response Title\\",
    \\"type\\": \\"webpage\\",
  }"
`;

exports[`<ManualInput /> should match baseline snapshot 1`] = `
<ManualInput
  store="__SKIPPED__"
>
  <div
    className="typeSelect"
  >
    <label
      className="label"
      htmlFor="type-select"
    >
      Citation Type
    </label>
    <select
      className="select"
      id="__SKIPPED__"
      onChange={[Function]}
      value="webpage"
    >
      <option
        aria-selected={false}
        value="bill"
      >
        Bill
      </option>
      <option
        aria-selected={false}
        value="book"
      >
        Book
      </option>
      <option
        aria-selected={false}
        value="chapter"
      >
        Book Section
      </option>
      <option
        aria-selected={false}
        value="legal_case"
      >
        Case
      </option>
      <option
        aria-selected={false}
        value="paper-conference"
      >
        Conference Proceeding
      </option>
      <option
        aria-selected={false}
        value="entry-encyclopedia"
      >
        Encyclopedia Entry
      </option>
      <option
        aria-selected={false}
        value="motion_picture"
      >
        Film
      </option>
      <option
        aria-selected={false}
        value="article"
      >
        Generic (Note)
      </option>
      <option
        aria-selected={false}
        value="speech"
      >
        Hearing
      </option>
      <option
        aria-selected={false}
        value="article-journal"
      >
        Journal Article
      </option>
      <option
        aria-selected={false}
        value="article-magazine"
      >
        Magazine Article
      </option>
      <option
        aria-selected={false}
        value="article-newspaper"
      >
        Newspaper Article
      </option>
      <option
        aria-selected={false}
        value="patent"
      >
        Patent
      </option>
      <option
        aria-selected={false}
        value="report"
      >
        Report
      </option>
      <option
        aria-selected={false}
        value="legislation"
      >
        Statute
      </option>
      <option
        aria-selected={false}
        value="broadcast"
      >
        Television Broadcast
      </option>
      <option
        aria-selected={false}
        value="thesis"
      >
        Thesis
      </option>
      <option
        aria-selected={true}
        value="webpage"
      >
        Web Page
      </option>
    </select>
  </div>
  <AutoCite
    getter={[Function]}
    kind="webpage"
    placeholder="URL"
  >
    <div
      className="autocite"
    >
      <label
        htmlFor="citequery"
      >
        Autocite
      </label>
      <input
        className="input"
        id="__SKIPPED__"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="URL"
        type="url"
        value=""
      />
      <Button
        disabled={true}
        flat={true}
        label="Search"
        onClick={[Function]}
        type="button"
      >
        <span>
          <button
            aria-label="Search"
            className="btn btnFlat"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Search
          </button>
        </span>
      </Button>
    </div>
  </AutoCite>
  <div
    className="scrollBoundaryAutocite"
    onWheel={[Function]}
  >
    <Callout
      intent="danger"
      onDismiss={[Function]}
    />
    <ContributorList
      citationType="webpage"
      people={Array []}
    >
      <h2
        className="heading"
      >
        Contributors
      </h2>
      <div
        className="btnRow"
      >
        <Button
          flat={true}
          label="Add contributor"
          onClick={[Function]}
          type="button"
        >
          <span>
            <button
              aria-label="Add contributor"
              className="btn btnFlat"
              onClick={[Function]}
              type="button"
            >
              Add contributor
            </button>
          </span>
        </Button>
      </div>
    </ContributorList>
    <MetaFields
      meta="__SKIPPED__"
    >
      <h2
        className="heading"
      >
        Web Page
      </h2>
      <div
        className="table"
      >
        <Field
          field={
            Object {
              "label": "Content Title",
              "required": true,
              "value": "title",
            }
          }
          meta="__SKIPPED__"
          onChange={[Function]}
        >
          <div
            className="row"
          >
            <label
              className="label"
              htmlFor="field-title"
            >
              Content Title
            </label>
            <input
              className="input"
              id="__SKIPPED__"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
          </div>
        </Field>
        <Field
          field={
            Object {
              "label": "Website Title",
              "required": true,
              "value": "container-title",
            }
          }
          meta="__SKIPPED__"
          onChange={[Function]}
        >
          <div
            className="row"
          >
            <label
              className="label"
              htmlFor="field-container-title"
            >
              Website Title
            </label>
            <input
              className="input"
              id="__SKIPPED__"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
          </div>
        </Field>
        <Field
          field={
            Object {
              "label": "URL",
              "required": true,
              "value": "URL",
            }
          }
          meta="__SKIPPED__"
          onChange={[Function]}
        >
          <div
            className="row"
          >
            <label
              className="label"
              htmlFor="field-URL"
            >
              URL
            </label>
            <input
              className="input"
              id="__SKIPPED__"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
          </div>
        </Field>
        <Field
          field={
            Object {
              "label": "Date",
              "pattern": "[0-2][0-9]{3}(?:(\\\\/(?:0[1-9]|1[0-2]))(\\\\/(?:[0-2][0-9]|3[0-1]))?)?",
              "title": "YYYY/MM/DD or YYYY/MM or YYYY",
              "value": "issued",
            }
          }
          meta="__SKIPPED__"
          onChange={[Function]}
        >
          <div
            className="row"
          >
            <label
              className="label"
              htmlFor="field-issued"
            >
              Date
            </label>
            <input
              className="input"
              id="__SKIPPED__"
              onChange={[Function]}
              pattern="[0-2][0-9]{3}(?:(\\\\/(?:0[1-9]|1[0-2]))(\\\\/(?:[0-2][0-9]|3[0-1]))?)?"
              title="YYYY/MM/DD or YYYY/MM or YYYY"
              type="text"
              value=""
            />
          </div>
        </Field>
        <Field
          field={
            Object {
              "label": "Date Accessed",
              "pattern": "[0-2][0-9]{3}(?:(\\\\/(?:0[1-9]|1[0-2]))(\\\\/(?:[0-2][0-9]|3[0-1]))?)?",
              "title": "YYYY/MM/DD or YYYY/MM or YYYY",
              "value": "accessed",
            }
          }
          meta="__SKIPPED__"
          onChange={[Function]}
        >
          <div
            className="row"
          >
            <label
              className="label"
              htmlFor="field-accessed"
            >
              Date Accessed
            </label>
            <input
              className="input"
              id="__SKIPPED__"
              onChange={[Function]}
              pattern="[0-2][0-9]{3}(?:(\\\\/(?:0[1-9]|1[0-2]))(\\\\/(?:[0-2][0-9]|3[0-1]))?)?"
              title="YYYY/MM/DD or YYYY/MM or YYYY"
              type="text"
              value=""
            />
          </div>
        </Field>
      </div>
    </MetaFields>
  </div>
</ManualInput>
`;

exports[`<ManualInput /> should not render autocite for anything but websites or books 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -12,11 +12,11 @@
      </label>
      <select
        className=\\"select\\"
        id=\\"__SKIPPED__\\"
        onChange={[Function res]}
-       value=\\"article\\"
+       value=\\"webpage\\"
      >
        <option
          aria-selected={false}
          value=\\"bill\\"
        >
@@ -57,11 +57,11 @@
          value=\\"motion_picture\\"
        >
          Film
        </option>
        <option
-         aria-selected={true}
+         aria-selected={false}
          value=\\"article\\"
        >
          Generic (Note)
        </option>
        <option
@@ -117,40 +117,114 @@
          value=\\"thesis\\"
        >
          Thesis
        </option>
        <option
-         aria-selected={false}
+         aria-selected={true}
          value=\\"webpage\\"
        >
          Web Page
        </option>
      </select>
    </div>
+   <AutoCite
+     getter={[Function anonymous]}
+     kind=\\"webpage\\"
+     placeholder=\\"URL\\"
+   >
      <div
-     className=\\"scrollBoundary\\"
+       className=\\"autocite\\"
+     >
+       <label
+         htmlFor=\\"citequery\\"
+       >
+         Autocite
+       </label>
+       <input
+         className=\\"input\\"
+         id=\\"__SKIPPED__\\"
+         onChange={[Function res]}
+         onKeyDown={[Function anonymous]}
+         placeholder=\\"URL\\"
+         type=\\"url\\"
+         value=\\"\\"
+       />
+       <Button
+         disabled={true}
+         flat={true}
+         label=\\"Search\\"
+         onClick={[Function res]}
+         type=\\"button\\"
+       >
+         <span>
+           <button
+             aria-label=\\"Search\\"
+             className=\\"btn btnFlat\\"
+             disabled={true}
+             onClick={[Function res]}
+             type=\\"button\\"
+           >
+             Search
+           </button>
+         </span>
+       </Button>
+     </div>
+   </AutoCite>
+   <div
+     className=\\"scrollBoundaryAutocite\\"
      onWheel={[Function anonymous]}
    >
      <Callout
        intent=\\"danger\\"
        onDismiss={[Function res]}
      />
+     <ContributorList
+       citationType=\\"webpage\\"
+       people={Array []}
+     >
+       <h2
+         className=\\"heading\\"
+       >
+         Contributors
+       </h2>
+       <div
+         className=\\"btnRow\\"
+       >
+         <Button
+           flat={true}
+           label=\\"Add contributor\\"
+           onClick={[Function res]}
+           type=\\"button\\"
+         >
+           <span>
+             <button
+               aria-label=\\"Add contributor\\"
+               className=\\"btn btnFlat\\"
+               onClick={[Function res]}
+               type=\\"button\\"
+             >
+               Add contributor
+             </button>
+           </span>
+         </Button>
+       </div>
+     </ContributorList>
      <MetaFields
        meta=\\"__SKIPPED__\\"
      >
        <h2
          className=\\"heading\\"
        >
-         Generic (Note)
+         Web Page
        </h2>
        <div
          className=\\"table\\"
        >
          <Field
            field={
              Object {
-               \\"label\\": \\"Text\\",
+               \\"label\\": \\"Content Title\\",
                \\"required\\": true,
                \\"value\\": \\"title\\",
              }
            }
            meta=\\"__SKIPPED__\\"
@@ -161,11 +235,11 @@
            >
              <label
                className=\\"label\\"
                htmlFor=\\"field-title\\"
              >
-               Text
+               Content Title
              </label>
              <input
                className=\\"input\\"
                id=\\"__SKIPPED__\\"
                onChange={[Function res]}
@@ -173,9 +247,133 @@
                type=\\"text\\"
                value=\\"\\"
              />
            </div>
          </Field>
+         <Field
+           field={
+             Object {
+               \\"label\\": \\"Website Title\\",
+               \\"required\\": true,
+               \\"value\\": \\"container-title\\",
+             }
+           }
+           meta=\\"__SKIPPED__\\"
+           onChange={[Function res]}
+         >
+           <div
+             className=\\"row\\"
+           >
+             <label
+               className=\\"label\\"
+               htmlFor=\\"field-container-title\\"
+             >
+               Website Title
+             </label>
+             <input
+               className=\\"input\\"
+               id=\\"__SKIPPED__\\"
+               onChange={[Function res]}
+               required={true}
+               type=\\"text\\"
+               value=\\"\\"
+             />
            </div>
+         </Field>
+         <Field
+           field={
+             Object {
+               \\"label\\": \\"URL\\",
+               \\"required\\": true,
+               \\"value\\": \\"URL\\",
+             }
+           }
+           meta=\\"__SKIPPED__\\"
+           onChange={[Function res]}
+         >
+           <div
+             className=\\"row\\"
+           >
+             <label
+               className=\\"label\\"
+               htmlFor=\\"field-URL\\"
+             >
+               URL
+             </label>
+             <input
+               className=\\"input\\"
+               id=\\"__SKIPPED__\\"
+               onChange={[Function res]}
+               required={true}
+               type=\\"text\\"
+               value=\\"\\"
+             />
+           </div>
+         </Field>
+         <Field
+           field={
+             Object {
+               \\"label\\": \\"Date\\",
+               \\"pattern\\": \\"[0-2][0-9]{3}(?:(\\\\\\\\/(?:0[1-9]|1[0-2]))(\\\\\\\\/(?:[0-2][0-9]|3[0-1]))?)?\\",
+               \\"title\\": \\"YYYY/MM/DD or YYYY/MM or YYYY\\",
+               \\"value\\": \\"issued\\",
+             }
+           }
+           meta=\\"__SKIPPED__\\"
+           onChange={[Function res]}
+         >
+           <div
+             className=\\"row\\"
+           >
+             <label
+               className=\\"label\\"
+               htmlFor=\\"field-issued\\"
+             >
+               Date
+             </label>
+             <input
+               className=\\"input\\"
+               id=\\"__SKIPPED__\\"
+               onChange={[Function res]}
+               pattern=\\"[0-2][0-9]{3}(?:(\\\\\\\\/(?:0[1-9]|1[0-2]))(\\\\\\\\/(?:[0-2][0-9]|3[0-1]))?)?\\"
+               title=\\"YYYY/MM/DD or YYYY/MM or YYYY\\"
+               type=\\"text\\"
+               value=\\"\\"
+             />
+           </div>
+         </Field>
+         <Field
+           field={
+             Object {
+               \\"label\\": \\"Date Accessed\\",
+               \\"pattern\\": \\"[0-2][0-9]{3}(?:(\\\\\\\\/(?:0[1-9]|1[0-2]))(\\\\\\\\/(?:[0-2][0-9]|3[0-1]))?)?\\",
+               \\"title\\": \\"YYYY/MM/DD or YYYY/MM or YYYY\\",
+               \\"value\\": \\"accessed\\",
+             }
+           }
+           meta=\\"__SKIPPED__\\"
+           onChange={[Function res]}
+         >
+           <div
+             className=\\"row\\"
+           >
+             <label
+               className=\\"label\\"
+               htmlFor=\\"field-accessed\\"
+             >
+               Date Accessed
+             </label>
+             <input
+               className=\\"input\\"
+               id=\\"__SKIPPED__\\"
+               onChange={[Function res]}
+               pattern=\\"[0-2][0-9]{3}(?:(\\\\\\\\/(?:0[1-9]|1[0-2]))(\\\\\\\\/(?:[0-2][0-9]|3[0-1]))?)?\\"
+               title=\\"YYYY/MM/DD or YYYY/MM or YYYY\\"
+               type=\\"text\\"
+               value=\\"\\"
+             />
+           </div>
+         </Field>
+       </div>
      </MetaFields>
    </div>
  </ManualInput>"
`;

exports[`<ManualInput /> should render autocite fields for books 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -12,20 +12,20 @@
      </label>
      <select
        className=\\"select\\"
        id=\\"__SKIPPED__\\"
        onChange={[Function res]}
-       value=\\"book\\"
+       value=\\"webpage\\"
      >
        <option
          aria-selected={false}
          value=\\"bill\\"
        >
          Bill
        </option>
        <option
-         aria-selected={true}
+         aria-selected={false}
          value=\\"book\\"
        >
          Book
        </option>
        <option
@@ -117,22 +117,21 @@
          value=\\"thesis\\"
        >
          Thesis
        </option>
        <option
-         aria-selected={false}
+         aria-selected={true}
          value=\\"webpage\\"
        >
          Web Page
        </option>
      </select>
    </div>
    <AutoCite
      getter={[Function anonymous]}
-     kind=\\"book\\"
-     pattern=\\"(?:[dxX]-?){10}|(?:[dxX]-?){13}\\"
-     placeholder=\\"ISBN\\"
+     kind=\\"webpage\\"
+     placeholder=\\"URL\\"
    >
      <div
        className=\\"autocite\\"
      >
        <label
@@ -143,13 +142,12 @@
        <input
          className=\\"input\\"
          id=\\"__SKIPPED__\\"
          onChange={[Function res]}
          onKeyDown={[Function anonymous]}
-         pattern=\\"(?:[dxX]-?){10}|(?:[dxX]-?){13}\\"
-         placeholder=\\"ISBN\\"
-         type=\\"text\\"
+         placeholder=\\"URL\\"
+         type=\\"url\\"
          value=\\"\\"
        />
        <Button
          disabled={true}
          flat={true}
@@ -178,11 +176,11 @@
      <Callout
        intent=\\"danger\\"
        onDismiss={[Function res]}
      />
      <ContributorList
-       citationType=\\"book\\"
+       citationType=\\"webpage\\"
        people={Array []}
      >
        <h2
          className=\\"heading\\"
        >
@@ -214,19 +212,19 @@
        meta=\\"__SKIPPED__\\"
      >
        <h2
          className=\\"heading\\"
        >
-         Book
+         Web Page
        </h2>
        <div
          className=\\"table\\"
        >
          <Field
            field={
              Object {
-               \\"label\\": \\"Title\\",
+               \\"label\\": \\"Content Title\\",
                \\"required\\": true,
                \\"value\\": \\"title\\",
              }
            }
            meta=\\"__SKIPPED__\\"
@@ -237,11 +235,11 @@
            >
              <label
                className=\\"label\\"
                htmlFor=\\"field-title\\"
              >
-               Title
+               Content Title
              </label>
              <input
                className=\\"input\\"
                id=\\"__SKIPPED__\\"
                onChange={[Function res]}
@@ -252,182 +250,56 @@
            </div>
          </Field>
          <Field
            field={
              Object {
-               \\"label\\": \\"Series Title\\",
-               \\"value\\": \\"collection-title\\",
+               \\"label\\": \\"Website Title\\",
+               \\"required\\": true,
+               \\"value\\": \\"container-title\\",
              }
            }
            meta=\\"__SKIPPED__\\"
            onChange={[Function res]}
          >
            <div
              className=\\"row\\"
            >
              <label
                className=\\"label\\"
-               htmlFor=\\"field-collection-title\\"
+               htmlFor=\\"field-container-title\\"
              >
-               Series Title
+               Website Title
              </label>
              <input
                className=\\"input\\"
                id=\\"__SKIPPED__\\"
                onChange={[Function res]}
+               required={true}
                type=\\"text\\"
                value=\\"\\"
              />
            </div>
          </Field>
          <Field
            field={
              Object {
-               \\"label\\": \\"Series Number\\",
-               \\"pattern\\": \\"[0-9]+\\",
-               \\"title\\": \\"One or more numbers, no spaces\\",
-               \\"value\\": \\"collection-number\\",
-             }
-           }
-           meta=\\"__SKIPPED__\\"
-           onChange={[Function res]}
-         >
-           <div
-             className=\\"row\\"
-           >
-             <label
-               className=\\"label\\"
-               htmlFor=\\"field-collection-number\\"
-             >
-               Series Number
-             </label>
-             <input
-               className=\\"input\\"
-               id=\\"__SKIPPED__\\"
-               onChange={[Function res]}
-               pattern=\\"[0-9]+\\"
-               title=\\"One or more numbers, no spaces\\"
-               type=\\"text\\"
-               value=\\"\\"
-             />
-           </div>
-         </Field>
-         <Field
-           field={
-             Object {
-               \\"label\\": \\"# of Pages\\",
-               \\"pattern\\": \\"[0-9]+\\",
-               \\"title\\": \\"One or more numbers, no spaces\\",
-               \\"value\\": \\"number-of-pages\\",
-             }
-           }
-           meta=\\"__SKIPPED__\\"
-           onChange={[Function res]}
-         >
-           <div
-             className=\\"row\\"
-           >
-             <label
-               className=\\"label\\"
-               htmlFor=\\"field-number-of-pages\\"
-             >
-               # of Pages
-             </label>
-             <input
-               className=\\"input\\"
-               id=\\"__SKIPPED__\\"
-               onChange={[Function res]}
-               pattern=\\"[0-9]+\\"
-               title=\\"One or more numbers, no spaces\\"
-               type=\\"text\\"
-               value=\\"\\"
-             />
-           </div>
-         </Field>
-         <Field
-           field={
-             Object {
-               \\"label\\": \\"Volume\\",
-               \\"pattern\\": \\"[0-9]+\\",
-               \\"title\\": \\"One or more numbers, no spaces\\",
-               \\"value\\": \\"volume\\",
-             }
-           }
-           meta=\\"__SKIPPED__\\"
-           onChange={[Function res]}
-         >
-           <div
-             className=\\"row\\"
-           >
-             <label
-               className=\\"label\\"
-               htmlFor=\\"field-volume\\"
-             >
-               Volume
-             </label>
-             <input
-               className=\\"input\\"
-               id=\\"__SKIPPED__\\"
-               onChange={[Function res]}
-               pattern=\\"[0-9]+\\"
-               title=\\"One or more numbers, no spaces\\"
-               type=\\"text\\"
-               value=\\"\\"
-             />
-           </div>
-         </Field>
-         <Field
-           field={
-             Object {
-               \\"label\\": \\"Edition\\",
-               \\"pattern\\": \\"[0-9]+\\",
-               \\"title\\": \\"One or more numbers, no spaces\\",
-               \\"value\\": \\"edition\\",
-             }
-           }
-           meta=\\"__SKIPPED__\\"
-           onChange={[Function res]}
-         >
-           <div
-             className=\\"row\\"
-           >
-             <label
-               className=\\"label\\"
-               htmlFor=\\"field-edition\\"
-             >
-               Edition
-             </label>
-             <input
-               className=\\"input\\"
-               id=\\"__SKIPPED__\\"
-               onChange={[Function res]}
-               pattern=\\"[0-9]+\\"
-               title=\\"One or more numbers, no spaces\\"
-               type=\\"text\\"
-               value=\\"\\"
-             />
-           </div>
-         </Field>
-         <Field
-           field={
-             Object {
-               \\"label\\": \\"Publisher\\",
+               \\"label\\": \\"URL\\",
                \\"required\\": true,
-               \\"value\\": \\"publisher\\",
+               \\"value\\": \\"URL\\",
              }
            }
            meta=\\"__SKIPPED__\\"
            onChange={[Function res]}
          >
            <div
              className=\\"row\\"
            >
              <label
                className=\\"label\\"
-               htmlFor=\\"field-publisher\\"
+               htmlFor=\\"field-URL\\"
              >
-               Publisher
+               URL
              </label>
              <input
                className=\\"input\\"
                id=\\"__SKIPPED__\\"
                onChange={[Function res]}
@@ -438,41 +310,12 @@
            </div>
          </Field>
          <Field
            field={
              Object {
-               \\"label\\": \\"Publisher Location\\",
-               \\"value\\": \\"publisher-place\\",
-             }
-           }
-           meta=\\"__SKIPPED__\\"
-           onChange={[Function res]}
-         >
-           <div
-             className=\\"row\\"
-           >
-             <label
-               className=\\"label\\"
-               htmlFor=\\"field-publisher-place\\"
-             >
-               Publisher Location
-             </label>
-             <input
-               className=\\"input\\"
-               id=\\"__SKIPPED__\\"
-               onChange={[Function res]}
-               type=\\"text\\"
-               value=\\"\\"
-             />
-           </div>
-         </Field>
-         <Field
-           field={
-             Object {
                \\"label\\": \\"Date\\",
                \\"pattern\\": \\"[0-2][0-9]{3}(?:(\\\\\\\\/(?:0[1-9]|1[0-2]))(\\\\\\\\/(?:[0-2][0-9]|3[0-1]))?)?\\",
-               \\"required\\": true,
                \\"title\\": \\"YYYY/MM/DD or YYYY/MM or YYYY\\",
                \\"value\\": \\"issued\\",
              }
            }
            meta=\\"__SKIPPED__\\"
@@ -490,11 +333,10 @@
              <input
                className=\\"input\\"
                id=\\"__SKIPPED__\\"
                onChange={[Function res]}
                pattern=\\"[0-2][0-9]{3}(?:(\\\\\\\\/(?:0[1-9]|1[0-2]))(\\\\\\\\/(?:[0-2][0-9]|3[0-1]))?)?\\"
-               required={true}
                title=\\"YYYY/MM/DD or YYYY/MM or YYYY\\"
                type=\\"text\\"
                value=\\"\\"
              />
            </div>"
`;
